cmd/ncproxy/service.go

```diff
@@ -0,0 +1,183 @@
+	s        *server
+	serveErr chan error
+}
+
+func getServicePath() (string, error) {
+	p, err := exec.LookPath(os.Args[0])
+	if err != nil {
+		return "", err
+	}
+	return filepath.Abs(p)
+}
+
+type serviceFailureActions struct {
+	ResetPeriod  uint32
+	RebootMsg    *uint16
+	Command      *uint16
+	ActionsCount uint32
+	Actions      uintptr
+}
+
+type scAction struct {
+	Type  uint32
+	Delay uint32
+}
+
+// See http://stackoverflow.com/questions/35151052/how-do-i-configure-failure-actions-of-a-windows-service-written-in-go
+const (
+	scActionNone    = 0
+	scActionRestart = 1
+
+	serviceConfigFailureActions = 2
+)
+
+func registerService() error {
+	p, err := getServicePath()
+	if err != nil {
+		return err
+	}
+	m, err := mgr.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = m.Disconnect()
+	}()
+
+	c := mgr.Config{
+		ServiceType:  windows.SERVICE_WIN32_OWN_PROCESS,
+		StartType:    mgr.StartAutomatic,
+		ErrorControl: mgr.ErrorNormal,
+		DisplayName:  "Ncproxy",
+		Description:  "Network configuration proxy",
+	}
+
+	// Configure the service to launch with the arguments that were just passed.
+	args := []string{"-run-service"}
+	for _, a := range os.Args[1:] {
+		if a != "-register-service" && a != "-unregister-service" {
+			args = append(args, a)
+		}
+	}
+
+	s, err := m.CreateService(serviceName, p, c, args...)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	t := []scAction{
+		{Type: scActionRestart, Delay: uint32(15 * time.Second / time.Millisecond)},
+		{Type: scActionRestart, Delay: uint32(15 * time.Second / time.Millisecond)},
+		{Type: scActionNone},
+	}
+	lpInfo := serviceFailureActions{ResetPeriod: uint32(24 * time.Hour / time.Second), ActionsCount: uint32(3), Actions: uintptr(unsafe.Pointer(&t[0]))}
+	return windows.ChangeServiceConfig2(s.Handle, serviceConfigFailureActions, (*byte)(unsafe.Pointer(&lpInfo)))
+}
+
+func unregisterService() error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = m.Disconnect()
+	}()
+
+	s, err := m.OpenService(serviceName)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	return s.Delete()
+}
+
+// Returns an indication to stop on successful SCM operation, and an error. For either register or unregister
+// return true so we can exit ncproxy and start it up as a Windows service.
+func registerUnregisterService() (bool, error) {
+	if *unregisterSvc {
+		if *registerSvc {
+			return true, errors.New("-register-service and -unregister-service cannot be used together")
+		}
+		return true, unregisterService()
+	}
+
+	if *registerSvc {
+		return true, registerService()
+	}
+
+	return false, nil
+}
+
+// launchService is the entry point for running ncproxy under SCM.
+func launchService(s *server, serveErr chan error) error {
+	if !*runSvc {
+		return nil
+	}
+
+	h := &handler{
+		fromsvc:  make(chan error),
+		s:        s,
+		serveErr: serveErr,
+	}
+
+	interactive, err := svc.IsAnInteractiveSession() // nolint:staticcheck
```

Comment:
what is this linter option for here?

Reply:
Leftover from the containerd implementation. Not sure, if needed, I'll remove and see

Reply:
`IsInteractiveSession` is deprecated in favor of `IsWindowsService`, that's why it was present.